### PR TITLE
2716241 Registration on login pane

### PR DESCRIPTION
**Pull request by haringsrob**
This is a proposal to have the registration form available on the login step.

I added an extra option to the pane config, so that the site builder can still choose not to show the form.

I also added the user_login hook, to redirect after the actual registration.

### CODE REVIEW

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php [hunk 1]
```diff
@@ -76,17 +77,23 @@ class Login extends CheckoutPaneBase implements CheckoutPaneInterface, Container
...
    *   The request stack.
    */
-  public function __construct(array $configuration, $plugin_id, $plugin_definition, CheckoutFlowInterface $checkout_flow, CredentialsCheckFloodInterface $credentials_check_flood, AccountInterface $current_user, EntityTypeManagerInterface $entity_type_manager, UserAuthInterface $user_auth, RequestStack $request_stack) {
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, CheckoutFlowInterface $checkout_flow, CredentialsCheckFloodInterface $credentials_check_flood, AccountInterface $current_user, EntityTypeManagerInterface $entity_type_manager, EntityManagerInterface $entity_manager, EntityFormBuilderInterface $entity_form_builder, UserAuthInterface $user_auth, RequestStack $request_stack) {
     parent::__construct($configuration, $plugin_id, $plugin_definition, $checkout_flow);
 
     $this->credentialsCheckFlood = $credentials_check_flood;
     $this->currentUser = $current_user;
     $this->entityTypeManager = $entity_type_manager;
+    $this->entityManager = $entity_manager;
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: Why inject the deprecated entityManager when you already have the non-deprecated entityTypeManager?

**Comment by bojanz** [comment]
Thanks for getting this rolling! The embedding of the entire existing registration form looks a bit messy, it would probably be better to copy over the AccountForm / RegistrationForm logic (but without the "needs admin approval" and the pass reset / preferred language stuff)

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php (hunk 1 above)
Comment (haringsrob) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: I will change this, after I created this, I worked on an issue about the same in core.. But I forgot I used it here.

**Comment by haringsrob** [comment]
Bojanz:

I wanted to do that, but I feel that that would not be the correct way to go. What if extra other modules modify the registration form? We would then make it impossible to use them? Or am I missing something?

**Comment by haringsrob** [comment]
Also, If we keep it in one form, and we would add required fields, they would become part of the same js validation (tried this first), when you then would submit the login form, you would be blocked by the registration fields.

**Comment by bojanz** [comment]
Both of these dilemmas are the same as for login.

You don't mark any field as required, then validate the fields if one of them was provided.
The context is sufficiently different that I'm not worried about the regular alter hooks not working (due to a different form ID).

**Comment by haringsrob** [comment]
@bojanz, if we do modify, then I assume the following fields are enough (no timezone etc,...):
schermafbeelding 2016-05-10 om 07 34 00

**Comment by haringsrob** [comment]
We do need an option to auto generate the username i think, as email + username is not user friendly.

**Comment by haringsrob** [comment]
@bojanz, updated the PR, waiting for feedback.

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php [hunk 2]
```diff
@@ -212,6 +236,29 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
...
+      '#title' => $this->t('Create new account'),
+      '#access' => !$this->configuration['allow_guest_checkout'] && $this->configuration['show_registration_form'],
+    ];
+    $pane_form['register']['mail'] = array(
+      '#type' => 'email',
+      '#title' => $this->t('Email address'),
+      '#description' => $this->t('A valid email address. All emails from the system will be sent to this address. The email address is not made public and will only be used if you wish to receive a new password or wish to receive certain news or notifications by email.'),
+      '#required' => FALSE,
+    );
+    $pane_form['register']['pass'] = array(
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: Long array usage.

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php [hunk 3]
```diff
@@ -212,6 +236,29 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
       '#op' => 'continue',
     ];
 
+    $pane_form['register'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('Create new account'),
+      '#access' => !$this->configuration['allow_guest_checkout'] && $this->configuration['show_registration_form'],
+    ];
+    $pane_form['register']['mail'] = array(
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: Long array usage.

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php [hunk 4]
```diff
@@ -220,53 +267,115 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
...
+          return;
+        }
 
-    $uid = $this->userAuth->authenticate($username, $password);
-    if (!$uid) {
-      $this->credentialsCheckFlood->register($this->clientIp, $username);
-      $form_state->setErrorByName('name', $this->t('Unrecognized username or password.'));
+        // Advanced validation Make sure the account does not exist yet. And
+        // that the username is unused/valid.
+        if ($this->entityTypeManager->getStorage('user')->loadByProperties(['mail' => $values['register']['mail']])) {
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: Why not get the $user_storage once then use it in both IFs?

File: modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php [hunk 5]
```diff
@@ -220,53 +267,115 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
...
+      case 'register':
+        $storage = $this->entityTypeManager->getStorage('user');
+        /** @var \Drupal\user\UserInterface $account */
+        $account = $storage->load($form_state->get('logged_in_uid'));
+        user_login_finalize($account);
+        $this->order->setOwner($account);
+        $this->credentialsCheckFlood->clearAccount($this->clientIp, $account->getAccountName());
+        break;
+
+      case 'continue':
```
Comment (bojanz) [inline modules/checkout/src/Plugin/Commerce/CheckoutPane/Login.php]: Why have both the continue and the default cases, which do the same thing? Kill one

**Comment by haringsrob** [comment]
I'm working on this.

File: modules/checkout/tests/src/Functional/CheckoutOrderTest.php [hunk 6]
```diff
@@ -96,4 +96,30 @@ public function testGuestOrderCheckout() {
...
+    $config->save();
+
+    $this->drupalLogout();
+    $this->drupalGet($this->product->toUrl()->toString());
+    $this->submitForm([], 'Add to cart');
+    $cart_link = $this->getSession()->getPage()->findLink('your cart');
+    $cart_link->click();
+    $this->submitForm([], 'Checkout');
+    $this->assertSession()->pageTextContains('Create new account');
+    $this->submitForm([
```
Comment (mglaman) [inline modules/checkout/tests/src/Functional/CheckoutOrderTest.php:110]: Can we also test validation? Existing email, invalid email

**Comment by haringsrob** [comment]
I think I have covered all the validations now.

**Pull request merged**